### ISSUE TITLE
V Drive Archive Suite

### ISSUE THREAD

**Issue by caseysmithpgh** [NYCPlanning/edm-overview#789]
Using Archive DTM as a model for moving old versions of ultimately all BytesProduction products to the V drive on a quarterly basis in coordination with PCNS approved V Drive write access.

**Comment by caseysmithpgh** [comment]
Re tested the script of Python 3.9.11 (sans ZIP64=True) and it was able to archive 4+ GB files successfully.

Now that script is working successfully, I am taking some time to restructure it into defined functions with the goal of separating out the archive and deletion sections. This will allow for clearer logging of archives/failed archives, and will condense what were previously 3 locations were successful archive test led to input file deletion into a single function.

Remaining items to address include ensuring empty year folders in bytes production are deleted once they are emptied.

**Comment by caseysmithpgh** [comment]
available_space() has a few logic problems involved with it. 
- While it is easy to calculate the size of the total items to be archived, this is not the figure needed to test whether we have enough space on the V:/. Instead we would need the summed size of the compressed files. We wont know this figure until the files are compressed (and the location where the compressed files will live is the V:/). 
- Estimating the compressed size could be an option -- but estimates of compression ratios vary quite a bit depending on file/folder size

**Comment by jackrosacker** [comment]
Yeah Matt and I chatted about this yesterday but I think you were deep in function writing mode at the time. 

The thinking was that since the calculated size is for the uncompressed files, it bakes a margin of error into the process. If the available space is pretty close to the uncompressed file size, then the compressed files will definitely fit. This is a bit of a crude safety margin, and probably worth getting a frame of reference for what the typical file size change is when running the compression tool. For example, if make_archive() makes the files significantly smaller, maybe it's worth re-working the test just because the V drive disk space seems to be so tightly administered. If it keeps the compressed files pretty close to the original size than the available_space() test is probably fine as-is.

Thoughts?

**Comment by caseysmithpgh** [comment]
@jackrosacker 
Right that makes sense. I've only tested the compression ration on Edes versions and not something larger-- seemed like the size reduced by <30% if I recall correctly--So not THAT significant. I suspect the ratio might be size dependent, so will have to give that a test.

If we went down that route, I would need to pick Hernan's brain about how to identify the available space allotted to the GISTEAM directory.

**Comment by caseysmithpgh** [comment]
TODO: 
- [x] Insert Deletion list logic via new defined function
- [x] Handle delete empty year folder logic being moved from major loop to delete_input function
- [x] add runtime to log
- [x] log number of items successfully archived
- [x] break out archive() function into 2 functions (compare archive to production and defining what needs to be archived | make archive)
- [x] build out analytics functions
    - [x] ID way to get size of .zip
- [scrap] explore logic that would reroute incorrectly located version to correct archive year (i.e. 20210131 in 2020 would become 20210131.zip in 2021)
- [ ] improve except error messages to include potential solutions/recommended troubleshooting
    - [ ] incorporate BREAK / or otherwise into Except statements to prevent script continuation and allow for clearer error msgs
- [x] confirm gdb/other contents aren't empty once zipped
- [x] calculate size of archivable BEFORE archive takes place in order to estimate allocation request for PCNS

NICE TO HAVES:
- [ ] sum amount of space of items to be archived and make sure that amount of space is available on V Drive

BEFORE RUNNING LIVE:
- [ ] ensure all products are un-commented
- [ ] swap root path variables

TEST STEPS:
- [x] Create folder i.e. 2020 with 2021 data inside
        RESULT: FAILS to reroute 2021 version to 2021 folder because its only told to locate it in mirrored parent folder (2020) 
- [x] Empty non year subfolder
        RESULT: DOES archive non year folder (which is fine w/matt) 
- [x] run all functions before delete > manually delete 2 zips > ensure result was logged correctly and no deletions of OG were made
        RESULT: successfully logs "failures", continues to attempt following deletions successfully
- [X] test archive year folder already existing (w/ duplicate content )
        RESULT: Overwrites .zip of the same version and continues on to next file
- [x] test origin vs archive file count difference
        RESULT: file was successfully removed from deletion_list, flag logged successfully 
- [ ] testing for lack of space on v drive
        Unable to stage scenario

**Comment by caseysmithpgh** [comment]
incorporated a PCNS_Allocation_Estimate jupyter notebook to inform the amount of space needed on V:/ to run archive script. It notably does not work well with the ini so it does have paths incorporated. Another downside is that the individual running the script needs to compare the data product lists in both scripts to ensure the estimate and archive are looking for the same products.

**Comment by caseysmithpgh** [comment]
Outstanding questions:

What are best practices when pulling any repo updates down to G:/ SCRIPTS (prod), specifically regarding the logs/ folder (not included in remote repo)? We don't want to erase the existing logs folder, so would a pull down to a secondary branch and subsequent merge-like action effectively incorporate existing logs into updated remote repo on production?

**Comment by caseysmithpgh** [comment]
Historical dof_dtm records were successfully archived and deleted from M:/

**Comment by caseysmithpgh** [comment]
Bytes archive was successfully running through the archive function, however it failed when it came across a file that was already zipped. this was because the code as it currently stood could not identify a zip file as an object, so i added language that would handle pre existing zip files at the version level. 

![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/d536d5f3-a969-439c-b59b-c15cdaeead23)

**Comment by caseysmithpgh** [comment]
Trouble with log merging ![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/af2cd9b7-a43a-4171-8011-e3c38775d355)

**Comment by fvankrieken** [comment]
I think that message can occur if the file is being used/edited when you attempt to stash. Beyond that, I'd be happy to help in a bit if you need any help sorting out git conflicts

**Comment by caseysmithpgh** [comment]
@fvankrieken ah that makes sense. I'll keep this in mind next time I'm handling this sort of thing--will let you know if I have continued troubles

**Comment by caseysmithpgh** [comment]
All qualified products were successfully archived on 3/28/2023. Only remaining step is for @croswell81 and @jackrosacker to run the script on local test folders to ensure documentation is fully comprehensive. Submitted request for Hernan revoke V:/ write access.

**Comment by caseysmithpgh** [comment]
@jackrosacker @croswell81 
Instructions for testing v-drive-archive. 

### Purpose: 
1. Ensure readme is comprehensive. Please notify me on any gaps in direction. 
2. Test pulling directory from Git into non G:/SCRIPTS directory and running from "local" development folder

### Step 1:
Copy the folders below from the X:/ drive to your C:/Temp folder (ensure capital "T") ![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/ca431563-5a0c-49fd-a00d-e9202e4eda6f)

Testing for regular unzipped version folders, zipped folders, loose files within year folders, and products not included in archive i.e. CPDB & ZAP:

![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/11e193df-e971-4fe8-9ab1-0d5c516e9527)
![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/afbc9436-5c60-4840-b4d3-bad42e80d611)
![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/23cca6da-f491-4480-94b1-963370540307)

### Step 2: 
Pull the v-drive-archive repo from git. Option 1: use git bash console. Option 2: Download zip archive + add /log sub folder ( no need to create logs, script should do this automatically). 

 ![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/0e0a8e79-42af-44ea-bf20-8362b412602d)

Save in your named folder here:
![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/9ea0688f-d26a-440d-9e6f-be8145adbdb5)
 
### Step 3:
Switch your config.ini paths to test paths 
![image.png](https://images.zenhubusercontent.com/628cede5d6c2595e02af266c/ae9e6474-4071-4d6f-8ab5-2c71c92e7915)

### Step 4: 
See README for instructions. Notify me of any gaps in documentation and I can provide additional guidance as needed.

**Comment by caseysmithpgh** [comment]
NOTE FOR MIH: 

MIH BP Distribution script errored because it was expecting 20190329 version to be in M:/GIS/BytesProd, however it was archived. To solve the issue I copied the version from V:/ back to M:/ (and unzipped). Hopefully we will have time to refactor MIH distribution scripts so this work around is no longer necessary--however if that is not the case, an exception can either be incorporated into the archive script for this version--or we will just need to replace the version in M:/ after running the archive script.

**Comment by jackrosacker** [comment]
> I copied the version from V:/ back to M:/ (and unzipped). Hopefully we will have time to refactor MIH distribution scripts so this work around is no longer necessary

Awesome. How would you like to handle ad hoc mid term script improvement items like this moving forward? 

The two tools that come to mind are repo-specific issue (MIH in this case), or adding to the existing script upgrade Zenhub issue. I think that adding as an MIH repo issue makes sense, and then we can summarize all pending Github issues when we decide to do a comprehensive repo review and refactor. Thoughts?

**Comment by jackrosacker** [comment]
Is there a way to bring an issue created on a specific repo into Zenhub? I would assume there is, but I've only ever kept them in their own silos. Stand up conversation!

**Comment by caseysmithpgh** [comment]
Jack provided [feedback](https://nyco365-my.sharepoint.com/:w:/g/personal/jrosacker_planning_nyc_gov/EdXp2o-6m6VPlHOX5cgJ39sBZG53IdqF15aZRYO_qyivyA?email=CSmith%40planning.nyc.gov&e=jfpMLE&CID=E44B949D-4EA5-4B38-ACC7-E9EA52A8B44F) last week. Most of it has been incorporated into the repo.
